### PR TITLE
PLANNER-736: Workbench: Prevent user from creating multiple @PlanningSolution data objects

### PR DESCRIPTION
**Pull request by mcimbora**
Part of:
https://github.com/droolsjbpm/guvnor/pull/416
https://github.com/droolsjbpm/jbpm-wb/pull/630
https://github.com/droolsjbpm/kie-wb-common/pull/681
https://github.com/droolsjbpm/optaplanner-wb/pull/134

### CODE REVIEW

File: optaplanner-wb-screens/optaplanner-wb-domain-editor/optaplanner-wb-domain-editor-client/src/main/java/org/optaplanner/workbench/screens/domaineditor/client/widgets/planner/PlannerDataObjectEditor.java [hunk 1]
```diff
@@ -121,7 +129,20 @@ protected void loadDataObject( DataObject dataObject ) {
...
         if ( dataObject != null ) {
             boolean hasPlanningEntity = dataObject.getAnnotation( PlannerDomainAnnotations.PLANNING_ENTITY_ANNOTATION ) != null;
-            boolean hasPlanningSolution = dataObject.getAnnotation( PlannerDomainAnnotations.PLANNING_SOLUTION_ANNOTATION ) != null;
+            boolean hasPlanningSolution = dataObject.getAnnotation( PLANNING_SOLUTION_ANNOTATION ) != null;
+
+            if ( !hasPlanningSolution ) {
+                dataModelerService.call( new RemoteCallback<List<Path>>() {
+                    @Override
+                    public void callback( List<Path> paths ) {
+                        view.enablePlanningSolutionCheckBox( paths.isEmpty() );
```
Comment (manstis) [inline optaplanner-wb-screens/optaplanner-wb-domain-editor/optaplanner-wb-domain-editor-client/src/main/java/org/optaplanner/workbench/screens/domaineditor/client/widgets/planner/PlannerDataObjectEditor.java]: Just an idea.. to disable these before making the async call.. if the call is, for whatever reason "slow", it gives the User the opportunity to click on something they aught not.

Comment (ge0ffrey) [reply optaplanner-wb-screens/optaplanner-wb-domain-editor/optaplanner-wb-domain-editor-client/src/main/java/org/optaplanner/workbench/screens/domaineditor/client/widgets/planner/PlannerDataObjectEditor.java]: @mcimbora while testing in development we're blind to latency. There are ways to simulate latency locally. IIRC on linux level, but maybe also through a firefox/chrome extension (which is probably easier to use). I recommend testing with a latency of 300ms per request (regardless of the size of the request, so bandwidth stays unlimited). This will immediately expose n+1 query problems, as those will have a latency of n times 300ms (= x minutes), resulting in very long, visible delays.

We need to do this kind of testing at least once in every Final release cycle (just like findbugs, running coverage, etc). 

@yurloc Maybe we can even automate it by adding latency in selenium tests? Not sure if this is possible.

**Review by manstis — APPROVED** [review]

File: optaplanner-wb-screens/optaplanner-wb-domain-editor/optaplanner-wb-domain-editor-client/src/main/java/org/optaplanner/workbench/screens/domaineditor/client/widgets/planner/PlannerDataObjectEditor.java (hunk 1 above)
Comment (yurloc) [reply optaplanner-wb-screens/optaplanner-wb-domain-editor/optaplanner-wb-domain-editor-client/src/main/java/org/optaplanner/workbench/screens/domaineditor/client/widgets/planner/PlannerDataObjectEditor.java]: @ge0ffrey The idea of using an artificial latency to expose n+1 query problems sounds interesting, CC'ing @jhrcek.

In this case, however, it wouldn't help because the current trend is to avoid using Selenium tests for checking validation, corner cases and negative testing (e.g. "given this and that, the checkbox should be disabled"). Due to maintenance cost we try to limit Selenium tests to happy path testing and end-to-end use cases.
